internal/core/metadata/operators/device_profile/value_descriptor.go

```diff
@@ -0,0 +1,196 @@
+
+		desc := contract.From(dr)
+		id, err := a.client.Add(&desc, a.ctx)
+		if err != nil {
+			a.logger.Error(fmt.Sprintf("Unable to create value descriptor: %s", err.Error()))
+			return err
+		}
+		a.logger.Debug(fmt.Sprintf("Created Value Descriptor id: %s", id))
+	}
+	return nil
+}
+
+// NewAddExecutor creates a new ValueDescriptorAddExecutor.
+func NewAddExecutor(ctx context.Context, client ValueDescriptorAdder, loggingClient logger.LoggingClient, drs ...contract.DeviceResource) ValueDescriptorAddExecutor {
+	return addValueDescriptor{
+		ctx:    ctx,
+		drs:    drs,
+		client: client,
+		logger: loggingClient,
+	}
+}
+
+// updateValueDescriptor encapsulates the data needed to update a value descriptor.
+type updateValueDescriptor struct {
+	ctx    context.Context
+	dp     contract.DeviceProfile
+	loader DeviceProfileLoader
+	client ValueDescriptorUpdater
+	logger logger.LoggingClient
+}
+
+// UpdateValueDescriptorExecutor updates a value descriptor.
+type UpdateValueDescriptorExecutor interface {
+	Execute() error
+}
+
+// Execute updates a value descriptor with the provided information.
+func (u updateValueDescriptor) Execute() error {
+	// Get pre-existing device profile so we can determine what to do with the device resources provided in the update. For example, update/create/delete.
+	persistedDeviceProfile, err := u.loader.GetDeviceProfileByName(u.dp.Name)
+	if err != nil {
+		return err
+	}
+
+	// Get names of all the device resources so we can check the valueDescriptorUsage with one call to Core-Data.
+	var persistedDeviceResourceNames []string
+	for _, persistedDeviceResource := range persistedDeviceProfile.DeviceResources {
+		persistedDeviceResourceNames = append(persistedDeviceResourceNames, persistedDeviceResource.Name)
+	}
+
+	// Check if any of the ValueDescriptors associated with the DeviceResources are in use.
+	// If so return an error stating all the ValueDescriptors which are in use.
+	valueDescriptorUsage, err := u.client.ValueDescriptorsUsage(persistedDeviceResourceNames, u.ctx)
+	if err != nil {
+		return err
+	}
+
+	var inUseValueDescriptors []string
+	for name, inUse := range valueDescriptorUsage {
+		if inUse {
+			inUseValueDescriptors = append(inUseValueDescriptors, name)
+		}
+	}
+
+	if len(inUseValueDescriptors) > 0 {
+		return errors.NewErrValueDescriptorsInUse(inUseValueDescriptors)
+	}
+
+	// Based on the DeviceProfile as it is before the update, determine which operation needs to be applied to each
+	// ValueDescriptor to get it in the desired state which is the information passed to the update command.
+	create, update, deleted := determineValueDescriptor(persistedDeviceProfile, u.dp)
+
+	// Execute the necessary operations to get the DeviceProfile to the desired state.
+	for _, d := range deleted {
+		err = u.client.DeleteByName(d.Name, u.ctx)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	for _, up := range update {
+		v, err := u.client.ValueDescriptorForName(up.Name, u.ctx)
```

Comment:
I really do not like doing this. If there is another way of doing this I would like to remove this. I thought about updating the `usage` endpoint in core-data to include the IDs in the response.